enhancements/20220125-concurrency-control-on-host-disk-operations.md

```diff
@@ -0,0 +1,271 @@
+# Concurrency Control on Host Disk Operations
+
+## Summary
+
+Some operations, such as formatting, might take a while for operating system to
+finish. It causes problems if the controller of host disk manager re-enters
+these criticial sections. To prevent re-entrance, additional disk management of
+Harvester needs a robust concurrency control on host disk operations.
+
+### Related Issues
+
+- Original bug report: https://github.com/harvester/harvester/issues/1718
+- Previous PR: https://github.com/harvester/node-disk-manager/pull/23
+
+## Motivation
+
+### Goals
+
+- Prevent criticial disk operations from re-entrances.
+- Control maximum count of concurrent disk operations within a time period.
+- Refactor the logic of blockdevice reconcilation to a more deterministic,
+  readable, and maintainable state.
+
+### Non-goals
+
+- Auto-recovering from disk operation failures is not in the scope. Node disk
+  manager does not know how to deal with this kind of failures (might due to 
+  hardware or OS failures). User should come in and tackle the issue manually
+  before re-triggering any futher operation.
+
+## Proposal
+
+This proposal is originally for tackling [issue #1718]. During the investigation
+of that issue, we found several potential flaws in the current implementation:
+
+1. After introduced [feature auto-provision], user is able to format and
+   provision a bunch of raw block device at once. And that leads to several 
```

Comment:
I don't understand why race conditions can happen here. Is each block device a separate CR? Kubernetes controller makes sure at any given time only one CR of the same name will be handled in the same process (which I assume will be one per node since block devices are local to the node). The queue is a built-in part of Kubernetes controller pattern.